Respect sharded dimensions when aten expaned/view consumes SymInt values

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99058
* #99040
* #98966

Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

mrshenli committed:
Respect sharded dimensions when aten expaned/view consumes SymInt values

Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

[ghstack-poisoned]

mrshenli committed:
Update on "Respect sharded dimensions when aten expaned/view consumes SymInt values"


Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

[ghstack-poisoned]

mrshenli committed:
Update on "Respect sharded dimensions when aten expaned/view consumes SymInt values"


Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

[ghstack-poisoned]

mrshenli committed:
Update on "Respect sharded dimensions when aten expaned/view consumes SymInt values"


Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

[ghstack-poisoned]

wanchaol commented on test/distributed/_spmd/test_tracing.py:
```diff
@@ -733,6 +733,25 @@ def train_step(mod, opt, inp):
+        opt = torch.optim.SGD(mod.parameters(), lr=0.01, foreach=True)
+        inp = torch.randn(2, 10).cuda(self.rank)
+        train_step(mod, opt, inp)
+        gm = train_step.__dict__[COMPILED_OBJECT_KEY].gm
```
hmmm should this simply be `train_step._compiled_obj.gm`?

wanchaol commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -159,6 +169,35 @@ def _remap_arg(node_to_obj: Dict[fx.Node, object], arg: object) -> object:
+
+
+# Dispatch override for
+SYM_INT_CONSUMERS: Dict[torch._ops.OpOverload, Callable] = {
```
would this approach also works for the tensor factory method (i.e. full, arange)?

mrshenli replied:
yep, I assume so.

wanchaol commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -491,7 +549,17 @@ def _convert_to_distributed(
+                else Replicate()
+            )
+            node_to_obj[node] = DSize(
+                size=dtensor._local_tensor.size(dim), placement=placement
```
nit: use `to_local` instead?

mrshenli replied:
I saw that calls into an Autograd function + a detach, so wanted to avoid those initially. But yep, let us use the public API from DTensor

wanchaol commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -159,6 +169,35 @@ def _remap_arg(node_to_obj: Dict[fx.Node, object], arg: object) -> object:
+    return DTensor.from_local(
+        local_tensor=op(args[0]._local_tensor, local_sizes),
+        device_mesh=args[0].device_mesh,
+        placements=[Shard(i) for i, p in enumerate(placements) if p.is_shard()],
```
is it only supporting `Shard`? but looks like line 180 handles Replicate() too?

mrshenli replied:
yep, only Shard, as we don't need special handling if SymInt represents a replicated dimension. Let me update line 180 to make it more clear.

wanchaol reviewed (approved):
stamp to unblock, have a couple of questions

mrshenli committed:
Update on "Respect sharded dimensions when aten expaned/view consumes SymInt values"


Currently, aten.expand always expands to the global dimension. Then, it
introduces additional slice and clone ops before running compute on
the expanded tensor with a local tensor.

In this commit, if we detect the op consumes a SymInt size, it respects
both local size and the dimension placements from where the SymInt was
extracted.

[ghstack-poisoned]

mrshenli commented:
Test failure is irrelevant

```
=========================== short test summary info ============================
FAILED [2.6467s] test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxWithOnnxRuntime_op_level_debug_True_dynamic_shapes_False::test_large_scale_exporter_with_tiny_gpt2 - AssertionError: Mixing fake modes NYI
============ 1 failed, 651 passed, 152 skipped in 103.51s (0:01:43) ============
Error: Process completed with exit code 1.
```

mrshenli commented:
@pytorchbot merge -f "onnx test failure is irrelevant"
